### PR TITLE
[WIP] Implement StableDiffusion3's MMDiT

### PR DESCRIPTION
**Pull request by kabachuha**
The goal of this PR is to provide an implementation of MMDiT (Multi-Modal Diffusion Transformer) by StabilityAI which processes not only the image latents in its transformer blocks, but also the text, and uses JointAttention on both embeddings.

See https://github.com/PKU-YuanGroup/Open-Sora-Plan/issues/43, https://stability.ai/news/stable-diffusion-3-research-paper / https://stabilityai-public-packages.s3.us-west-2.amazonaws.com/Stable+Diffusion+3+Paper.pdf for more details and the model diagram

Tasks:
- [x] Implement the MMDiT block
- [x] Implement Joint Attention
- [x] Support for mixed CLIP and T5 embeddings + pooled embeddings timestep projection
- [x] Correct attention to be similar to the rest of the codebase
- [x] Compose the upper-level Transformer structure
- [x] Modify Latte to have MMDiT-like architecture as well
- [ ] Make a test run and hope it won't crash

### CODE REVIEW

**Comment by oahzxl** [comment]
thanks for your contribution. we will try to merge this pr when sd3 is released

File: opendit/modules/joint_attn.py [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+        return qkv
+    
+    def split_qkv(self, qkv, B, N, total_N, num_heads):
+        if self.sequence_parallel_type == "ulysses":
+            q, k, v = qkv.split(self.head_dim * self.num_heads, dim=-1)
+            q = all_to_all_comm(q, self.sequence_parallel_group)
+            k = all_to_all_comm(k, self.sequence_parallel_group)
+            v = all_to_all_comm(v, self.sequence_parallel_group)
+
+            if self.enable_flashattn:
```
Comment (MaximilienLC) [inline opendit/modules/joint_attn.py:176]: `self.enable_flashattn` -> `self.attention_mode == "flash"` ? See also line 200, 206

File: opendit/modules/joint_attn.py [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+
+        return self.scale * x_normed
+
+class JointAttention(nn.Module):
+    def __init__(self, txt_dim, pix_dim,
+                        num_heads=8,
+                        qkv_bias=False,
+                        attn_drop=0.,
+                        proj_drop=0.,
+                        use_lora=False,
```
Comment (MaximilienLC) [inline opendit/modules/joint_attn.py:82]: unused

File: opendit/modules/joint_attn.py [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+                        attention_mode='math',
+                        eps=1e-12,
+                        causal=True,
```
Comment (MaximilienLC) [inline opendit/modules/joint_attn.py:85]: unused

**Comment by pzpzpzp2** [comment]
@oahzxl 
> thanks for your contribution. we will try to merge this pr when sd3 is released

if sd3 is released doesn't this pr become redundant?

**Comment by kabachuha** [comment]
Closing because of soon SD3 release, and because I have a better idea of how to employ this architecture for a drastically new model

Thanks Y-all for reviews!

**Pull request closed without merging**